File: chapter 08/recipe 8.10/DniScQueuedErrorsVB/DniScQCErrorObj.vb
Imports System.IO
Imports System.Runtime.InteropServices
Imports System.EnterpriseServices

<ClassInterface(ClassInterfaceType.None)> _
Public Class DniScQCErrorObj
    Inherits ServicedComponent
    Implements IQueuedComponent
    Implements IPlaybackControl

    Public Sub LogMessage(ByVal message As String) _
        Implements IQueuedComponent.LogMessage

        'called if the message cannot be delivered
        'to the original component

        Dim writer As StreamWriter _
            = New StreamWriter("c:\QCErrorLog.txt", True)
        writer.WriteLine(message)
        writer.Flush()
        writer.Close()
    End Sub

    Public Sub FinalClientRetry() _
        Implements IPlaybackControl.FinalClientRetry

    End Sub

    Public Sub FinalServerRetry() _
        Implements IPlaybackControl.FinalServerRetry

    End Sub
End Class

File: chapter 08/recipe 8.10/DniScQueuedErrors/DniScQCErrorObj.cs
using System;
using System.IO;
using System.Runtime.InteropServices;
using System.EnterpriseServices;

namespace DniScQueuedComponent
{
	[ClassInterface(ClassInterfaceType.None)]
	public class DniScQCErrorObj : ServicedComponent, 
		  IPlaybackControl, IQueuedComponent
	{
		public void LogMessage(string message)
		{
			//called if the message cannot be delivered
			//to the original component

			using (StreamWriter writer
				= new StreamWriter(@"c:\QCErrorLog.txt", true))
			{
				writer.WriteLine(message);
				writer.Flush();
			}
		}

		public void FinalClientRetry()
		{
			//notification of a delivery failure
			//on the client side. 
		}

		public void FinalServerRetry()
		{
			//notification of a playback failure
			//on the server
		}
	}
}
